File: backend/tiggle-root/tiggle/src/main/java/com/side/tiggle/global/auth/OAuth2Attribute.java
package com.side.tiggle.global.auth;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;

import java.time.LocalDate;
import java.util.HashMap;
import java.util.Map;

@Builder(access = AccessLevel.PRIVATE)
@Getter
public class OAuth2Attribute {
    private Map<String, Object> attributes;
    private String attributeKey;
    private String email;
    private String profileUrl;
    private String birthString;
    private String nickname;
    private String providerId;

    static OAuth2Attribute of(String provider, String attributeKey,
                              Map<String, Object> attributes) {
        switch (provider) {
            case "google":
                return ofGoogle(attributeKey, attributes);
            case "kakao":
                return ofKakao(attributeKey, attributes);
            case "naver":
                return ofNaver(attributeKey, attributes);
            default:
                throw new IllegalStateException();
        }
    }

    private static OAuth2Attribute ofGoogle(String attributeKey,
                                            Map<String, Object> attributes) {
        return OAuth2Attribute.builder()
                .email((String) attributes.get("email"))
                .profileUrl((String)attributes.get("picture"))
                .nickname((String)attributes.get("name"))
                .providerId((String)attributes.get("sub"))
                .attributes(attributes)
                .attributeKey((String)attributes.get(attributeKey))
                .build();
    }

    private static OAuth2Attribute ofNaver(String userNameAttributeName, Map<String, Object> attributes) {
        Map<String, Object> response = (Map<String, Object>)attributes.get("response");

        return OAuth2Attribute.builder()
                .nickname((String) response.get("name"))
                .email((String) response.get("email"))
                .profileUrl((String) response.get("profile_image"))
                .providerId((String) response.get("id"))
                .attributes(attributes)
                .attributeKey(userNameAttributeName)
                .build();
    }

    private static OAuth2Attribute ofKakao(String userNameAttributeName, Map<String, Object> attributes) {
        Map<String, Object> kakaoAccount = (Map<String, Object>)attributes.get("kakao_account");
        Map<String, Object> kakaoProfile = (Map<String, Object>)kakaoAccount.get("profile");

        return OAuth2Attribute.builder()
                .nickname((String) kakaoProfile.get("nickname"))
                .email((String) kakaoAccount.get("email"))
                .profileUrl((String) kakaoProfile.get("profile_image_url"))
                .providerId(attributes.get("id").toString())
                .attributes(attributes)
                .attributeKey(userNameAttributeName)
                .build();
    }

    Map<String, Object> toMap() {
        Map<String, Object> map = new HashMap<>();
        map.put("id", attributeKey);
        map.put("key", attributeKey);
        map.put("email", email);
        map.put("profile_url", profileUrl);
        map.put("nickname", nickname);

        return map;
    }
}

File: backend/tiggle-root/tiggle/src/main/kotlin/com/side/tiggle/global/auth/OAuth2Attribute.kt
package com.side.tiggle.global.auth

class OAuth2Attribute(
    val attributes: Map<String, Any>,
    val attributeKey: String,
    val email: String,
    val profileUrl: String,
    val nickname: String,
    val providerId: String,
    val birthString: String? = null
    ) {

    fun toMap(): Map<String, Any> {
        return mapOf(
            "id" to attributeKey,
            "key" to attributeKey,
            "email" to email,
            "profile_url" to profileUrl,
            "nickname" to nickname
        )
    }

    companion object {
        fun of(
            provider: String,
            attributeKey: String,
            attributes: Map<String, Any>
        ): OAuth2Attribute {
            return when(provider) {
                "google" -> ofGoogle(attributeKey, attributes)
                "naver" -> ofNaver(attributeKey, attributes)
                "kakao" -> ofKakao(attributeKey, attributes)
                else -> throw IllegalStateException()
            }
        }

        fun ofGoogle(
            attributeKey: String,
            attributes: Map<String, Any>
        ): OAuth2Attribute {
            return OAuth2Attribute(
                email = attributes["email"] as String,
                profileUrl = attributes["picture"] as String,
                nickname = attributes["name"] as String,
                providerId = attributes["sub"] as String,
                attributes = attributes,
                attributeKey = attributeKey
            )
        }

        fun ofNaver(
            userNameAttributeName: String,
            attributes: Map<String, Any>
        ): OAuth2Attribute {
            val response = attributes["response"] as Map<String, Any>
            return OAuth2Attribute(
                email = response["email"] as String,
                nickname = response["name"] as String,
                profileUrl = response["profile_image"] as String,
                providerId = response["id"] as String,
                attributes = attributes,
                attributeKey = userNameAttributeName
            )
        }

        fun ofKakao(
            userNameAttributeName: String,
            attributes: Map<String, Any>
        ): OAuth2Attribute {
            val kakaoAccount = attributes["kakao_account"] as Map<String, Any>
            val kakaoProfile = attributes["profile"] as Map<String, Any>
            return OAuth2Attribute(
                email = kakaoAccount["email"] as String,
                nickname = kakaoProfile["nickname"] as String,
                profileUrl = kakaoProfile["profile_image_url"] as String,
                providerId = attributes["id"] as String,
                attributes = attributes,
                attributeKey = userNameAttributeName
            )
        }
    }
}